fix: handle repeatable attributes

ph-fritsche opened the pull request:
Introduces `isTransient` on `AttributeDriver` for better handling of repeatable attributes. Cleans up `AttributeReader` API a little.

Closes #8755

ph-fritsche committed:
fix: handle repeatable attributes

greg0ire commented on lib/Doctrine/ORM/Mapping/Driver/AttributeDriver.php:
```diff
@@ -38,6 +40,23 @@ public function __construct(array $paths)
+        $classAnnotations = $this->reader->getClassAnnotations(new ReflectionClass($className));
+
+        foreach ($classAnnotations as $a) {
+            $annot = is_array($a) ? $a[0] : $a;
```
@ph-fritsche can `$a` really be an array?

ph-fritsche replied:
```php
            $annot = /* is_array($a) ? $a[0] : */ $a;
```
results in `phpunit` reporting:
```
There was 1 error:

1) Doctrine\Tests\ORM\Mapping\AttributeDriverTest::testIsTransientWithRepeatableAttributes
TypeError: get_class(): Argument #1 ($object) must be of type object, array given

/workspaces/misc/doctrine-orm/lib/Doctrine/ORM/Mapping/Driver/AttributeDriver.php:52
/workspaces/misc/doctrine-orm/tests/Doctrine/Tests/ORM/Mapping/AttributeDriverTest.php:85
```
due to:
https://github.com/doctrine/orm/blob/332c314e9f055d729555f2861fcb81ee50e482b5/lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php#L26-L28
https://github.com/doctrine/orm/blob/332c314e9f055d729555f2861fcb81ee50e482b5/lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php#L66-L85

greg0ire replied:
Indeed. The phpdoc of `convertToAttributeInstances` is wrong, this is why the build fails.

greg0ire replied:
Also, the phpdoc of `Doctrine\Common\Annotations\Reader` is wrong, and that's in another package :grimacing:

greg0ire replied:
Or rather `AttributeReader` does not implement the interface it says it implements. Widening the return type in the interface wouldn't be a BC break for classes implementing it, but it would be a BC break for calling code. Not sure what the best course of action would be here @beberlei . I think we should widen it in `doctrine/annotations`, but to be clean, that would require a major release :grimacing: or since it's just phpdoc, we pray it does not break to many static analysis builds and release it as a minor?

ph-fritsche replied:
Maybe `Doctrine\ORM\AttributeDriver` should not extend `Doctrine\Persistence\Mapping\Driver\AnnotationDriver` at all.
How about we internally refactor `Doctrine\Persistence\Mapping\Driver\AnnotationDriver::getAllClassNames` into another class that could be reused by `Doctrine\ORM\AttributeDriver`?
This way `doctrine/persistence` and `doctrine/annotation` would not have any breaking changes.

If you want to be sentimental about the version increment on this package, one could argue that the declaration of `Doctrine\ORM\AttributeDriver` was broken from the beginning in `2.9` so breaking that inheritance here might be released as a fix.

beberlei replied:
That is a course of action we could take for the next minor release. To fix this bug the small solution plus suppression of tooling errors is my preference.

greg0ire replied:
I just thought of another quite simple solution that does not imply changing other packages: let's use an iterable object instead of an array to represent a bug of attributes. As a bonus it will make the code clearer and will answer `can $a really be an array?` because you will have `$a instanceof RepeatableAttributeCollection` instead of `is_array($a)`. WDYT?

greg0ire replied:
Here is what it would look like: https://github.com/ph-fritsche/doctrine-orm/pull/1

greg0ire committed:
Restore interface compatibility

A reader is supposed to only ever return objects. By introducing
RepeatableAttributeCollection, we fulfill the interface and improve
clarity.

ph-fritsche committed:
Merge pull request #1 from greg0ire/repeatable-attribute-collection

Restore interface compatibility

ph-fritsche committed:
refactor: accurate type declarations and returns

ph-fritsche committed:
refactor: remove unused use

beberlei commented:
Can you exclude AttributeReader from php-stan and psalm analysis for now? We should upgrade to check with 8.0 in 2.10 branch.

beberlei committed:
Ignore AttributeReader in phpstan and psalm to pass on CI running PHP 7.4.

beberlei commented:
@ph-fritsche just thought the ping pong fixing some config changes might going to be tedious and made the change myself for now. Lets see what the CI says about this.

beberlei reviewed (dismissed):
This is good for now to fix the error.

I would say we could improve this in the future (2.10) by changing `AttributeReader` to have only three functions `getAnnotations() : array<Annotation>`, `getFilteredAnnotations(string $className) : array<Annotation>` and `getAnnotation(string $className) : Annotation` which throws when the class name is repeatable.

SenseException commented on tests/Doctrine/Tests/ORM/Mapping/AttributeDriverTest.php:
```diff
@@ -78,6 +78,13 @@ public function testOriginallyNestedAttributesDeclaredWithoutOriginalParent(): v
         $this->assertEquals(['assoz_id', 'assoz_id'], $metadata->associationMappings['assoc']['joinTableColumns']);
     }
+
+    public function testIsTransientWithRepeatableAttributes(): void
+    {
+        $driver      = $this->loadDriver();
+        $isTransient = $driver->isTransient(AttributeEntityStartingWithRepeatableAttributes::class);
+        $this->assertFalse($isTransient);
+    }
```
Covering the `true` case too would be nice.

ph-fritsche committed:
test: isTransient
